extensions/amp-ad-network-adzerk-impl/0.1/amp-ad-network-adzerk-impl.js

```diff
@@ -0,0 +1,272 @@
+    super(element);
+
+    /** @private {?CreativeMetaDataDef} */
+    this.creativeMetaData_ = null;
+  };
+
+  /**
+   * Validate the tag parameters.  If invalid, ad ad will not be displayed.
+   * @override
+   */
+  isValidElement() {
+    if (!this.win['DOMParser']) {
+      dev().error(TAG, 'Missing DOM Parser');
+      return false;
+    }
+    return !!this.getAdUrl();
+  }
+
+  /** @override */
+  getSigningServiceNames() {
+    // Does not utilize crypto signature based AMP creative validation.
+    // TODO(keithwrightbos): move import of crypto validation into
+    // implementations, reducing adzerk binary size.
+    return [];
+  }
+
+  /** @override */
+  getAdUrl() {
+    const src = this.element.getAttribute('src');
+    if (!/^https:\/\/adzerk.com\?id=\d+$/i.test(src)) {
+      return '';
+    }
+    if (getMode(this.win).localDev) {
+      return `http://ads.localhost:${this.win.location.port}` +
+        '/adzerk/' + /^https:\/\/adzerk.com\?id=(\d+)/.exec(src)[1];
+    }
+    // TODO(adzerk): specify expected src path.
+    return /^https:\/\/adzerk.com\?id=\d+$/i.test(src) ? src : '';
+  }
+
+  /**
+   * Fetch and parse template from AMP cache.  Result is stored in global in
+   * order to reduce overhead when template is used multiple times.
+   * @param {number} templateId
+   * @return {!Promise<!CachedTemplateDef>}
+   * @private
+   */
+  retrieveTemplate_(templateId) {
+    // Retrieve template from AMP cache.
+    TemplateCache[templateId] = TemplateCache[templateId] ||
+        Services.xhrFor(this.win)
+            .fetchText(getMode(this.win).localDev ?
+              `http://ads.localhost:${this.win.location.port}` +
+                `/a4a_template/adzerk/${templateId}` :
+              `${urls.cdn}/c/s/adzerk/${templateId}`,
+            TEMPLATE_CORS_CONFIG)
+            .then(response => response.text())
+            .then(template => this.parseTemplate_(template));
+    TemplateCache[templateId].access = Date.now();
+    const cacheKeys = /**@type {!Array<number>}*/(Object.keys(TemplateCache));
+    if (cacheKeys.length > 5) {
+      dev().warn(TAG, 'Trimming template cache');
+      // Evict oldest entry to ensure memory usage is minimized.
+      cacheKeys.sort((a, b) =>
+        TemplateCache[b].access - TemplateCache[a].access);
+      delete TemplateCache[cacheKeys[cacheKeys.length - 1]];
+    }
+    dev().assert(TemplateCache[templateId]);
+    return TemplateCache[templateId];
+  }
+
+  /**
+   * Extracts metadata from template head used for preferential render.
+   * @param {string} template
+   * @return {!CachedTemplateDef}
+   * @private
+   */
+  parseTemplate_(template) {
+    mustacheParse(template);
+    // TODO(keithwrightbos): this is temporary until AMP cache can be modified
+    // to parse and supply this information as opposed to requiring client
+    // side parsing.  For now build metadata and remove portions of template.
+    // Only the body will be used for mustache expansion.
```

Comment:
does it make sense to wrap the content with `<template>`, like what we do for amp-mustache:

```html
<body>
  Some static content.

  <template type="amp-mustache">
    dynamic content goes here {{var}}
  </template>
<body>
```

Reply:
In this way, you might be able to prevent the usage of `new DOMParser`. Instead you can use `document.importNode`

Reply:
Hmmm... that's an interesting idea and it would ensure that elements that utilize scripts (e.g. amp-animation) would not be removed.  Will change...

Reply:
I'm not really sure how `<template>` or `importNode` helps us here. Can someone offer some more details? How, for example, does it ensure that elements utilizing scripts do not get removed?

Reply:
document.importNode requires a node as an argument, not a string. I'll keep DOMParser for now, unless someone has a suggestion.